shorebird_code_push/lib/src/shorebird_updater_io.dart

```diff
@@ -0,0 +1,177 @@
+typedef IsolateRun = Future<R> Function<R>(
+  FutureOr<R> Function(), {
+  String? debugName,
+});
+
+/// {@template shorebird_updater_io}
+/// The Shorebird IO Updater.
+/// {@endtemplate}
+class ShorebirdUpdaterImpl implements ShorebirdUpdater {
+  /// {@macro shorebird_updater_io}
+  ShorebirdUpdaterImpl(this._updater, {IsolateRun? run})
+      : _run = run ?? Isolate.run {
+    try {
+      // If the Shorebird Engine is not available, this will throw an exception.
+      _updater.currentPatchNumber();
+      _isAvailable = true;
+    } catch (_) {
+      logShorebirdEngineUnavailableMessage();
+      _isAvailable = false;
+    }
+  }
+
+  late final bool _isAvailable;
+
+  final Updater _updater;
+
+  final IsolateRun _run;
+
+  @override
+  bool get isAvailable => _isAvailable;
+
+  @override
+  Future<Patch?> readCurrentPatch() async {
+    if (!_isAvailable) return null;
+
+    return _run(
+      () {
+        try {
+          final patchNumber = _updater.currentPatchNumber();
+          return patchNumber > 0 ? Patch(number: patchNumber) : null;
+        } catch (error) {
+          throw ReadPatchException(message: '$error');
+        }
+      },
+    );
+  }
+
+  @override
+  Future<Patch?> readNextPatch() async {
+    if (!_isAvailable) return null;
+
+    return _run(
+      () {
+        try {
+          final patchNumber = _updater.nextPatchNumber();
+          return patchNumber > 0 ? Patch(number: patchNumber) : null;
+        } catch (error) {
+          throw ReadPatchException(message: '$error');
+        }
+      },
+    );
+  }
+
+  @override
+  Future<UpdateStatus> checkForUpdate() async {
+    if (!_isAvailable) return UpdateStatus.unavailable;
+
+    final isUpdateAvailable = await _run(_updater.checkForUpdate);
+    if (isUpdateAvailable) return UpdateStatus.outdated;
+
+    final (current, next) = await (readCurrentPatch(), readNextPatch()).wait;
+    return next != null && current?.number != next.number
+        ? UpdateStatus.restartRequired
+        : UpdateStatus.upToDate;
+  }
+
+  @override
+  Future<void> update() async {
+    if (!_isAvailable) return;
+
+    Pointer<UpdateResult> result = nullptr;
+
+    try {
+      result = await _run(_updater.update);
+    } catch (_) {
+      return _legacyFallback();
+    }
+
+    try {
+      if (result == nullptr) {
+        const reason = UpdateFailureReason.unknown;
+        final message = reason.toFailureMessage();
+        throw UpdateException(reason: reason, message: message);
+      }
+
+      final status = result.ref.status;
+
+      if (status == SHOREBIRD_UPDATE_INSTALLED) return;
+
+      final reason = status.toFailureReason();
+      final details = result.ref.message != nullptr
+          ? result.ref.message.cast<Utf8>().toDartString()
+          : 'unknown';
+      final message = reason.toFailureMessage(details);
+      throw UpdateException(message: message, reason: reason);
+    } finally {
+      _updater.freeUpdateResult(result);
+    }
```

Comment:
Does this whole try belong nested inside the first one (e.g. with a helper function) and then the finally belong on the first try?

Reply:
What are you concerned about? The first try/catch throwing while still somehow having allocated a result? I'm not sure how that would be possible.

Reply:
As it's currently written I think it's fine.  If we ever add anything else to that first try block, we could end up leaking, since really what this is trying to clean up is the allocation from within the first try block.

Reply:
We shouldn’t add anything else to the first try block since the whole point of the first try is to handle exceptions when calling update 🤷